### DISCUSSION TITLE
[Thinking out loud] - Dynamic/interpolated/concatenated classes?

### DISCUSSION THREAD

**Discussion by VicGUTT** [tailwindlabs/tailwindcss#5985] [Ideas]
I'm pumped by this new feature! Awesome work!
The on demand feature and arbitrary values support got me excited for a possible way to allow concatenated classes with Tailwind.

We all know currently with Tailwind **it is important to avoid dynamically creating class strings in our templates with string concatenation, otherwise PurgeCSS won't know to preserve those classes.** _[Docs](https://tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html)_

And on the other hand, even with the new JIT mode, classes like `bg-$myVariable` _(PHP)_ and `bg-${myVariable}` _(JS)_ are evaluated at runtime and Tailwind will never be able to know what the resulting class will be.

That said, we could create a new syntax: `prefix-{token}` or `prefix-{*}`, that would instruct Tailwind to generate all possible values for that _prefix + { + token + }_ combo, example:
```html
<button
    class="
        mt-{spacing} or mt-{*}
        text-{colors} or text-{*}
        bg-{colors}-500 or bg-{*}-500

        mt-$foo
        text-$bar
        bg-$baz-500
    "
>

<!--
    Would generate:
        - .mt-px .mt-0 .mt-0.5 .mt-1 ... .mt-96
        - text-white .text-indigo-50 ... .text-indigo-900 ... .text-blue-50 ... .text-blue-900
        - .bg-indigo-500 .bg-pink-500 ... .bg-gray-500
-->
```

But I'm not convinced by this idea, and it might even be out of scope for the JIT feature.

On the other hand, and in the spirit of the on demand classes with arbitrary values like `w-[calc(100%+1rem)]` I think we could introduce CSS custom properties: `w-[var(--myVariable,fallback)]` or better, **Tailwind custom properties**: `w-[--tw-w-4]` which would allow us to get as close as possible to dynamic classes.

Usage example:

```html
<button class="bg-[var(--myVariable,2rem)] w-[--tw-w-4]">

<!--
    Would generate:
        - .bg-\[var\(--myVariable\,2rem\)\] {
            background-color: var(--myVariable, 2rem);
        }
        - .w-\[--tw-w-4\] {
            width: 1rem;
            or
            width: var(--tw-w-4, 1rem);
        }
-->
```

Like said in the issue title, I am currently just thinking out loud. And with the features this new JIT brings, being able to handle dynamically constructed classes is probably the last remaining _annoying_ thing with/limitation of Tailwind.

This issue is meant as a "food for thought" idea and would be better suited for Github's Discussions feature if it was available for this repository.

**Comment by yellow1912** [comment]
I wonder if it somehow aligns with your idea but how about a build client running on the browser (worker if necessarily). This can be super helpful for website builder apps. When we drop in a new element, add new class then new rules can be appended on the fly. We could run a build on the server but I think the delay even in seconds can be unbearable.

**Comment by VicGUTT** [comment]
While I agree the idea is interesting it is definitely out of scope for this projet (at least at this point). I'd love for the idea of using Tailwind's JIT on the browser to be explored at some point though, if only to bring back devtools autocompletion.

**Comment by benjaminpreiss** [comment]
@VicGUTT I love the idea! Nevertheless, I think it would be better to have something like runcss as a runtime complement. That way we can keep the syntax?

**Comment by VicGUTT** [comment]
I did not know of RunCSS but the idea is indeed interesting and I love out of the box thinking like this. I don't think I would use a similar approach in production though as I would not want my users to pay any byte/latency/CPU etc... penalty. And if I were, I would most likely use the ["Tailwindcss JIT CDN"](https://github.com/beyondcode/tailwindcss-jit-cdn)

**Comment by benjaminpreiss** [comment]
Yeah, @VicGUTT I feel you... I didn't know about the CDN, so that might be a valid alternative. Nevertheless, I think a little extra CPU is a small price for keeping the syntax uniform. What happens if tailwind changes the underlying CSS for a class? Then we have to go search in the style tags for the custom CSS we added to substitute a tailwind class

**Comment by VicGUTT** [comment]
I'm not sure I'm following. What do you mean by "_keeping the syntax uniform_" ? If Tailwind ever changes the underlying rules referenced by a class that would probably mean we made a conscious choice to jump a major version _(from v2.x.x to v3.0.0 etc...)_

**Comment by benjaminpreiss** [comment]
@VicGUTT True!

**Comment by tmcgann-neotax** [comment]
Perhaps there's a better place to continue this discussion. If there is point me in the right direction...

Would it be possible to create a plugin for webpack or for a transpiler like Babel that parses your code at build-time, recognizes dynamic class names, identifies if the interpolated values are static and deterministic, and then includes the corresponding class names in your bundle?

Even if it is possible, it raises the question of whether or not this is a good idea. This could be a total foot-gun that creates a lot of confusion because the deterministic class names are included but the non-deterministic ones aren't (because they are determined at runtime). As a result, developers may not realize the dynamic values they are supplying are non-deterministic at build-time and wonder why their code/CSS isn't working the way they expect.

The most common reason I want to use string interpolation is because I want my components that use tailwind classes to be configurable at build-time. I don't want to have to enumerate all the possible tailwind class names, since this defeats the point of not including the ones I don't use anyway. Things like `space-x-${value}` might be configurable on a `Flexbox` or `Stack` component that's designed to be a reusable container that groups children and handles their flow and spacing. While scenarios in which you would want this to be configurable at runtime are conceivable, I don't think they necessarily represent the majority. Regardless of usage, they are separate use cases.

A pure runtime solution might be the best way to go. Or maybe a hybrid of both if it makes sense, adds enough value to do something like this.
